### PR TITLE
PIPRES-199: Subscription customer address update

### PR DESCRIPTION
**Pull request by mandan2**
Implemented two hooks in order to update customer addresses for subscription orders. Used trait to have access to property in the same runtime (didn't want to fill mollie.php anymore). How the flow works:

- At first original subscription order has a link with customer address. After we update this address, new address gets created and the old one gets "deleted" flag. We insert newly created address ID into mol_recurring_order table to avoid updating original subscription order data.
- If we try to update newly created address, this address won't get flagged as deleted and will just be updated in the same instance, so our handler won't be triggered. This happens as newly created address was not linked to any of the existing orders ir ps_orders table. Same address ID keeps being used so our flow is not needed.

In mol_recurring_order table created two ned address id columns to save most up to date addresses there, without changing original order data.
Also created decorator for original PrestaShop order presenter lazy array as we wanted to show addresses from mol_recurring_order order, not from original subscription order.

### CODE REVIEW

File: mollie.php [hunk 1]
```diff
@@ -33,6 +33,8 @@
 
 class Mollie extends PaymentModule
 {
+    use Mollie\Subscription\Traits\HookTraits;
```
Comment (margud) [inline mollie.php:36]: import namespace?

File: subscription/Decorator/RecurringOrderLazyArray.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * @arrayAccess
+     *
+     * @return array
+     */
+    public function getAddresses(): array
+    {
+        $order = $this->order;
+
+        $orderAddresses = [
```
Comment (margud) [inline subscription/Decorator/RecurringOrderLazyArray.php:49]: i guess you could create object here and not array but optional

Comment (mandan2) [reply subscription/Decorator/RecurringOrderLazyArray.php:49]: I must use this same pattern as original OrderLazyArray uses array with those specific keys.

File: subscription/Repository/RecurringOrderRepository.php [hunk 3]
```diff
@@ -24,4 +26,23 @@ public function findAllBy(array $keyValueCriteria): ?PrestaShopCollection
 
         return $result;
     }
+
+    public function getCustomerOrdersByOldAddress(int $customerId, int $oldAddressId): ?array
```
Comment (margud) [inline subscription/Repository/RecurringOrderRepository.php:30]: are you using this function?

Comment (mandan2) [reply subscription/Repository/RecurringOrderRepository.php:30]: Nice, it is not used, forgot to delete.

**Review by margud — APPROVED** [review]

**Pull request merged**